wen-restart/src/wen_restart.rs

```diff
@@ -281,18 +301,127 @@ pub(crate) fn find_heaviest_fork(
-            expected_parent = slot;
+            if !block_meta.is_full() {
+                return Err(WenRestartError::BlockNotFull(*slot).into());
+            }
+            expected_parent = *slot;
         } else {
-            return Err(WenRestartError::BlockNotFound(slot).into());
+            return Err(WenRestartError::BlockNotFound(*slot).into());
+        }
+    }
+    let heaviest_fork_bankhash = find_bankhash_of_heaviest_fork(
+        heaviest_fork_slot,
+        slots,
+        blockstore.clone(),
+        bank_forks.clone(),
+        root_bank,
+        &exit,
+    )?;
+    info!(
+        "Heaviest fork found: slot: {}, bankhash: {:?}",
+        heaviest_fork_slot, heaviest_fork_bankhash
+    );
+    Ok((heaviest_fork_slot, heaviest_fork_bankhash))
+}
+
+fn find_bankhash_of_heaviest_fork(
+    heaviest_fork_slot: Slot,
+    slots: Vec<Slot>,
+    blockstore: Arc<Blockstore>,
+    bank_forks: Arc<RwLock<BankForks>>,
+    root_bank: Arc<Bank>,
+    exit: &Arc<AtomicBool>,
+) -> Result<Hash> {
+    // Find the hash of the heaviest fork, if block hasn't been replayed, replay to get the hash.
+    let heaviest_fork_bankhash = bank_forks
+        .read()
+        .unwrap()
+        .get(heaviest_fork_slot)
+        .map(|bank| bank.hash());
+    if let Some(hash) = heaviest_fork_bankhash {
+        return Ok(hash);
+    }
+
+    let leader_schedule_cache = LeaderScheduleCache::new_from_bank(&root_bank);
+    let replay_tx_thread_pool = rayon::ThreadPoolBuilder::new()
+        .thread_name(|i| format!("solReplayTx{i:02}"))
+        .build()
+        .expect("new rayon threadpool");
+    let recyclers = VerifyRecyclers::default();
+    let mut timing = ExecuteTimings::default();
+    let opts = ProcessOptions::default();
+    // Grab one write lock until end of function because we are the only one touching bankforks now.
+    let mut my_bankforks = bank_forks.write().unwrap();
+    // Check that the existing banks are frozen and link to the expected parent.
+    let mut parent_slot = root_bank.slot();
+    for slot in &slots {
+        if exit.load(Ordering::Relaxed) {
+            return Err(WenRestartError::Exiting.into());
         }
+        if let Some(bank) = my_bankforks.get(*slot) {
+            if !bank.is_frozen() {
+                return Err(WenRestartError::BlockNotFrozenAfterReplay(*slot, None).into());
+            }
+            if bank.parent_slot() != parent_slot {
+                return Err(WenRestartError::BlockNotLinkedToExpectedParent(
+                    *slot,
+                    Some(bank.parent_slot()),
+                    parent_slot,
+                )
+                .into());
+            }
+        }
+        parent_slot = *slot;
     }
-    Ok((expected_parent, Hash::default()))
+    // Now replay all the missing blocks.
+    let mut parent_bank = root_bank;
+    for slot in slots {
+        if exit.load(Ordering::Relaxed) {
+            return Err(WenRestartError::Exiting.into());
+        }
```

Comment:
who would set this early exit here? seems like we'll always play to completion

Reply:
Impatient user could Ctrl+C at any time.

Reply:
Agreed with leaving the check for exit flag - some other thread could hit an issue (ie ABS)